io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -178,29 +167,17 @@ sealed abstract class IO[E, A] { self =>
    * The error parameter of the returned `IO` may be chosen arbitrarily, since
    * it is guaranteed the `IO` action does not raise any errors.
    */
-  final def attempt[E2]: IO[E2, E \/ A] = (self.tag: @switch) match {
-    case IO.Tags.Point =>
-      val io = self.asInstanceOf[IO.Point[E, A]]
-
-      new IO.Point(() => \/-(io.value()))
-
-    case IO.Tags.Strict =>
-      val io = self.asInstanceOf[IO.Strict[E, A]]
-
-      new IO.Strict(\/-(io.value))
-
-    case IO.Tags.SyncEffect =>
-      val io = self.asInstanceOf[IO.SyncEffect[E, A]]
-
-      new IO.SyncEffect(() => \/-(io.effect()))
-
-    case IO.Tags.Fail =>
-      val io = self.asInstanceOf[IO.Fail[E, A]]
+  final def attempt[E2, B](err: E => IO[E2, B])(succ: A => IO[E2, B]): IO[E2, B] =
+    (self.tag: @switch) match {
+      case IO.Tags.Fail =>
+        val io = self.asInstanceOf[IO.Fail[E, A]]
+        err(io.error)
 
-      new IO.Strict(-\/(io.error))
+      case _ => new IO.Attempt(self, err, succ)
+    }
 
-    case _ => new IO.Attempt(self)
-  }
+  final def attemptEither[E2]: IO[E2, Either[E, A]] =
```

Comment:
We may want to rename `attempt` -> `catchAll`, and `attemptEither` -> `attempt`.

Reply:
We already have a `catchAll` which acts only on the error branch and I think better deserves that name. What about `flatCatchMap` or `flatBimap`? :laughing:

Reply:
I see Monix calls it `redeem`. Maybe we should just go with that.